Sources/Differentiation/Zip+Differentiable.swift

```diff
@@ -0,0 +1,167 @@
+#if canImport(_Differentiation)
+
+import _Differentiation
+
+@derivative(of: zip)
+@inlinable
+public func _vjpZip<Sequence1, Sequence2>(
+    _ sequence1: Sequence1, _ sequence2: Sequence2
+) -> (
+    value: Zip2Sequence<Sequence1, Sequence2>,
+    pullback: (Zip2Sequence<Sequence1, Sequence2>.TangentVector) -> (Sequence1.TangentVector, Sequence2.TangentVector)
+) where
+    Sequence1: Differentiable & RangeReplaceableCollection,
+    Sequence2: Differentiable & RangeReplaceableCollection,
+    Sequence1.Element: Differentiable,
+    Sequence2.Element: Differentiable,
+    Sequence1.TangentVector: RangeReplaceableCollection,
+    Sequence2.TangentVector: RangeReplaceableCollection,
+    Sequence1.TangentVector.Element == Sequence1.Element.TangentVector,
+    Sequence2.TangentVector.Element == Sequence2.Element.TangentVector
+{
+    (
+        value: zip(sequence1, sequence2),
+        pullback: { v in
+            (v.sequence1, v.sequence2)
+        }
+    )
+}
+
+extension Zip2Sequence: @retroactive Differentiable where
+    Sequence1: Differentiable & RangeReplaceableCollection,
+    Sequence2: Differentiable & RangeReplaceableCollection,
+    Sequence1.Element: Differentiable,
+    Sequence2.Element: Differentiable,
+    Sequence1.TangentVector: RangeReplaceableCollection,
+    Sequence2.TangentVector: RangeReplaceableCollection,
+    Sequence1.TangentVector.Element == Sequence1.Element.TangentVector,
+    Sequence2.TangentVector.Element == Sequence2.Element.TangentVector
+{
+    @inlinable
+    public mutating func move(by offset: TangentVector) {
+        // TODO: inside the stdlib we might be able to do this work in place
+        var result1 = Sequence1()
+        var result2 = Sequence2()
+
+        for (original, offset) in zip(self, offset) {
+            var original = original
+
+            original.0.move(by: offset.0)
+            original.1.move(by: offset.1)
+
+            result1.append(original.0)
+            result2.append(original.1)
+        }
+        self = zip(result1, result2)
+    }
+
+    @inlinable
+    public func differentiableMap<Result: Differentiable>(_ transform: @differentiable(reverse) (Sequence1.Element, Sequence2.Element)
+        -> Result
+    ) -> [Result] {
+        self.map(transform)
+    }
+
+    @derivative(of: differentiableMap)
+    @inlinable
+    public func _vjpDifferentiableMap<Result: Differentiable>(_ transform: @differentiable(reverse) (Sequence1.Element, Sequence2.Element)
+        -> Result
+    ) -> (value: [Result], pullback: ([Result].TangentVector) -> Zip2Sequence.TangentVector) {
+        var results: [Result] = []
+        // TODO: reserveCapacity
```

Comment:
Maybe document an explanation of why the TODO, or what is required before this would be possible?

Reply:
I for now reservedCapacity to underEstimatedCount which should work when using arrays. Thanks for pointing out